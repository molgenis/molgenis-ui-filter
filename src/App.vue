<template>
  <div
    id="app"
    class="container"
  >
    <div class="row">
      <div class="col col-md-3">
        <filter-container
          v-model="selections"
          :filters="filters"
          :filters-shown="filtersShown"
          :can-edit="true"
          @update="updateState"
        />
      </div>
      <div class="col col-md-8">
        <active-filters
          v-model="selections"
          :filters="filters"
        />
        <br><br>
        <pre><strong>selections:</strong>{{ selections }}</pre>
        <pre><strong>visibility:</strong>{{ filtersShown }}</pre>
      </div>
    </div>
  </div>
</template>

<script>
import Vue from 'vue'
import * as components from './components/filters'
import { FilterContainer, ActiveFilters } from './components/'

import data from '../tests/mockdata'

// API setup for testing filters.
import axios from 'axios'

const api = axios.create({
  baseURL: 'http://localhost:8080/api/',
  timeout: 15000
})

export default Vue.extend({
  name: 'App',
  'components': { FilterContainer, ActiveFilters, ...components },
  data () {
    return {
      selections: {
        search: 'test'
      },
      filtersShown: ['multi-filter', 'search', 'color', 'name', 'age', 'number', 'checkbox-options'],
      filters: [
        {
          name: 'multi-filter',
          label: 'Filter with multiple options',
          collapsed: false,
          entity: 'root_hospital_diagnosis',
          metaQuery: async (entity) => api.get(`metadata/${entity}`),
          dataQuery: async (query, entity, metadata) => {
            const nameAttr = metadata.data.data.attributes.items.filter((i) => i.data.labelAttribute).map((i) => i.data.name)[0]
            const data = await api.get(`data/${entity}?q=${nameAttr}=like=${query}`)
            return data.data.items.map((i) => {
              return { value: i.data.id, text: i.data[nameAttr] }
            })
          },
          options: (checked, entity) => api.get(`data/${entity}?q=id=in=(${checked.join(',')})`),
          type: 'multi-filter'
        },
        {
          name: 'search',
          label: 'Search',
          description: 'search by string',
          placeholder: 'placeholder',
          type: 'string-filter',
          collapsable: false
        },
        {
          name: 'color',
          label: 'Color',
          collapsed: false,
          bulkOperation: true,
          options: () => {
            return new Promise(
              function (resolve) {
                resolve(data.checkboxOptions)
              })
          },
          type: 'checkbox-filter'
        },
        {
          name: 'number',
          label: 'Number',
          collapsed: false,
          type: 'number-filter'
        },
        {
          name: 'age',
          label: 'Age',
          collapsed: false,
          min: -10,
          max: 10,
          step: 0.01,
          type: 'range-filter'
        },
        {
          name: 'name',
          label: 'Name',
          description: 'Name of object',
          type: 'string-filter',
          collapsed: false
        },
        {
          name: 'string',
          label: 'String',
          description: 'search by string',
          placeholder: 'placeholder',
          type: 'string-filter',
          collapsable: true,
          collapsed: false
        },
        {
          name: 'checkbox',
          label: 'Checkbox',
          collapsed: false,
          bulkOperation: true,
          maxVisibleOptions: 1,
          options: () => {
            return new Promise(
              function (resolve) {
                resolve([{ value: 'red', text: 'Red' }, { value: 'green', text: 'Green' }, { value: 'blue', text: 'Blue' }])
              })
          },
          type: 'checkbox-filter'
        },
        {
          name: 'checkbox-options',
          label: 'Checkbox lots of options',
          collapsed: false,
          bulkOperation: true,
          maxVisibleOptions: 1,
          options: () => {
            return new Promise(
              function (resolve) {
                resolve([{ value: 'red', text: 'Red' }, { value: 'green', text: 'Green' }, { value: 'blue', text: 'Blue' }])
              })
          },
          type: 'checkbox-filter'
        },
        {
          name: 'long-name',
          label: 'Way too long name to really fit in the user interface',
          collapsed: true,
          bulkOperation: true,
          options: () => {
            return new Promise(
              function (resolve) {
                resolve([{ value: 'no', text: 'No' }, { value: 'yes', text: 'Yes' }])
              })
          },
          type: 'checkbox-filter'
        }
      ]
    }
  },
  methods: {
    updateState (newState) {
      this.filtersShown = newState
    }
  }
})
</script>

<style>
</style>
