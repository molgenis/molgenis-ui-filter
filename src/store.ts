import axios from 'axios'
import data from '../tests/mockdata'

const api = axios.create({
  baseURL: 'http://localhost:8080/api/',
  timeout: 15000
})

// @ts-ignore
api.entity = 'root_hospital_diagnosis'

const providers = {
  multiFilter: async function (query:any) {
    // @ts-ignore
    const metadata = await api.get(`metadata/${api.entity}`)
    const nameAttr = metadata.data.data.attributes.items.filter((i:any) => i.data.labelAttribute).map((i) => i.data.name)[0]
    // @ts-ignore
    const data = await api.get(`data/${api.entity}?q=${nameAttr}=like=${query}`)
    return data.data.items.map((i:any) => {
      return { id: i.data.id, name: i.data[nameAttr] }
    })
  }
}

const widget = {
  collapsed: true,
  collapsible: true
}

const filterState = {
  filters: [
    {
      id: 'multi-filter',
      label: 'Multi filter',
      maxVisibleOptions: 5,
      search: '',
      options: [],
      provider: providers.multiFilter,
      type: 'MultiFilter',
      selection: [], // Maps to b-form-checkbox-group model
      active: true
    },
    {
      id: 'search',
      label: 'Search',
      description: 'search by string',
      placeholder: 'placeholder',
      type: 'StringFilter',
      selection: 'test',
      active: false
    },
    {
      id: 'color',
      label: 'Color',
      bulkOperation: true,
      maxVisibleOptions: 5,
      options: [],
      provider: () => {
        return new Promise((resolve) => {
          resolve(data.checkboxOptions)
        })
      },
      type: 'checkbox-filter',
      selection: [],
      active: false
    },
    {
      id: 'number',
      label: 'Number',
      type: 'NumberFilter',
      min: Number.MIN_SAFE_INTEGER,
      max: Number.MAX_SAFE_INTEGER,
      useSlider: false,
      step: 1,
      selection: null,
      active: false
    },
    {
      id: 'age',
      label: 'Age',
      min: -10,
      max: 10,
      step: 0.01,
      type: 'RangeFilter',
      useSlider: true,
      selection: [null, null],
      active: false
    },
    {
      id: 'name',
      label: 'Name',
      description: 'Name of object',
      type: 'StringFilter',
      selection: null,
      active: false
    },
    {
      id: 'string',
      label: 'String',
      description: 'search by string',
      placeholder: 'placeholder',
      type: 'StringFilter',
      selection: null,
      active: false
    },
    {
      id: 'checkbox',
      label: 'Checkbox',
      bulkOperation: true,
      maxVisibleOptions: 1,
      options: [],
      provider: () => {
        return new Promise((resolve) => {
          resolve(data.checkboxOptions)
        })
      },
      type: 'CheckboxFilter',
      selection: [],
      active: false
    },
    {
      id: 'checkbox-options',
      label: 'Checkbox lots of options',
      bulkOperation: true,
      maxVisibleOptions: 1,
      options: [],
      provider: () => {
        return new Promise((resolve) => {
          resolve(data.checkboxOptions)
        })
      },
      type: 'CheckboxFilter',
      selection: null,
      active: false
    },
    {
      id: 'long-name',
      bulkOperation: true,
      label: 'Way too long name to really fit in the user interface',
      options: [],
      provider: () => {
        return new Promise(
          function (resolve) {
            resolve([{ id: 'no', name: 'No' }, { id: 'yes', name: 'Yes' }])
          })
      },
      type: 'CheckboxFilter',
      selection: null,
      active: false
    }
  ]
}

for (const filter of filterState.filters) {
  Object.assign(filter, widget)
}

export default filterState
