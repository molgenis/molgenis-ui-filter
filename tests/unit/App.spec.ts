import { shallowMount, createLocalVue } from '@vue/test-utils'
import App from '@/App.vue'
import BootstrapVue from 'bootstrap-vue'
import store from '@/store'

const VueStash = require('vue-stash')

const localVue = createLocalVue()
localVue.use(VueStash.default)

describe('StringFilter.vue', () => {
  const wrapper = shallowMount(App, {
    localVue,
    parentComponent: { data: () => ({ store }) }
  })

  it('matches the snapshot', () => {
    expect(wrapper.element).toMatchSnapshot()
  })
})
