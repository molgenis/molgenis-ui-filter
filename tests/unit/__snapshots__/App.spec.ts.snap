// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StringFilter.vue matches the snapshot 1`] = `
<div
  class="container"
  id="app"
>
  <div
    class="row"
  >
    <div
      class="col col-md-3"
    >
      <filter-container-stub
        canedit="true"
        filters="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]"
        filtersshown="disease,search,color,name,age,number,checkbox-options"
        value="[object Object]"
      />
    </div>
     
    <div
      class="col col-md-8"
    >
      <active-filters-stub
        filters="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]"
        value="[object Object]"
      />
       
      <br />
      <br />
       
      <pre>
        <strong>
          selections:
        </strong>
        {
  "search": "test"
}
      </pre>
       
      <pre>
        <strong>
          visibility:
        </strong>
        [
  "disease",
  "search",
  "color",
  "name",
  "age",
  "number",
  "checkbox-options"
]
      </pre>
    </div>
  </div>
</div>
`;
